tutorials/cloud-functions-async/index.md

```diff
@@ -0,0 +1,150 @@
+
+When a caller makes a request of a service, it is asking it perform some work. Depending on the use-case, there are basically three expectations of this caller:
+
+### Synchronous Request/Response
+![sync](image/sync-request.png)
+
+The caller will wait until the work is done, expecting a result. This pattern can be directly and simply handled by [Cloud Functions](https://cloud.google.com/functions/) with an [HTTP trigger](https://cloud.google.com/functions/docs/calling/http).
+
+### Asynchronous Work Queue
+![work-queue](image/work-queue.png)
+
+The caller does not need to wait for the work to be done, and does not need to followup on the completion status.
+
+There are a couple ways this can be solved on Google Cloud Platform in a serverless way. You can use Cloud PubSub patterns for [long running tasks](https://cloud.google.com/solutions/using-cloud-pub-sub-long-running-tasks) or you can use a dedicated service with [Cloud Tasks](https://cloud.google.com/tasks/).
+
+### Asynchronous Stateful Jobs
+![](image/stateful-job.png)
+The caller does not want to wait for the work to be done, but would like to be able to inquire as to the completion of the request.
+
+This pattern is common enough that Google has defined a standard of a [long-running-operation API contract](https://github.com/googleapis/googleapis/tree/master/google/longrunning) used in multiple APIs.
+
+This is a basic pattern which may have some use-case specific details.
+
+## Serverless Stateful Jobs
+
+This tutorial uses several managed services to implement this pattern including [PubSub](https://cloud.google.com/pubsub/) and [Firestore](https://cloud.google.com/firestore/).
+
+![](image/arch.png)
+
+You can implement job state storage or a work queue with alternatives like Cloud SQL and Cloud Tasks. This tutorial uses the support of [golang](https://golang.org/) in Cloud Functions, but any supported language can be used to fulfill the pattern.
+
+### Job resource definition
+
+A simple custom job definition is defined:
+
+```
+type JobState int
+
+const (
+	Created JobState = iota
+	Running
+	Completed
+	Failed
+)
+
+type Job struct {
+	ID         string    `json:"id"`
+	CreateTime time.Time `json:"created-time"`
+	DoneTime   time.Time `json:"done-time" firestore:"DoneTime,omitempty"`
+	Done       bool
+	Result     string
+	State      JobState
+	Task       map[string]interface{} `json:"-"`
+	// can add a requester, source IP etc if needed
+}
+```
+
+When a job request is received, it is given an ID and the details of the work are included in a task field. This task is sent into the work queue as PubSub payload, with the Job ID as an [PubSub attribute](https://cloud.google.com/pubsub/docs/publisher#custom-attributes). When a worker picks up the task, it moves the job from `Created` -> `Running`. When the task is complete, it moves from `Running` to `Completed` (success) or `Failed` and writes any result back into the Firestore document.
```

Comment:
PubSub payload -> a Cloud PubSub payload

with the Job ID as an [PubSub attribute] -> with the Job ID as a Cloud [PubSub attribute]

`Created` -> `Running` -> `Created` to `Running`

Reply:
done